R/dataone.R

```diff
@@ -0,0 +1,73 @@
+
+
+
+
+
+library(dataone)
+library(datapack)
+library(mime)
+
+
+dataone_node <- function(){
+  if(!is.null(getOption("dataone_test_token")))
+    return( dataone::D1Client("STAGING2", "urn:node:mnTestKNB") )
+  dataone::D1Client("PROD", "urn:node:KNB")
+}
+
+resolve_dataone <- function(id, url_only = FALSE){
+  d1c <-  dataone_node()
+  paste0(d1c@cn@baseURL, "/v2/resolve/", utils::URLencode(id, TRUE))
+}
+
+## Create a data object with content-based id and a sha-256 checksum
+data_object <- function(file, format = mime::guess_type(file), ...){
+  library(datapack)
+  hash <- paste0(openssl::sha256(file(file)))
+  id <- paste0("hash://sha256/", hash)
+  d1Object <- new("DataObject", id, format=mime::guess_type(file), filename=file, ...)
+  
+  ## Won't be needed once this is the new default, see https://github.com/DataONEorg/rdataone/issues/257
+  d1Object@sysmeta@checksum <- gsub("^hash://\\w+/", "", id)
+  d1Object@sysmeta@checksumAlgorithm <- "SHA-256"
+  d1Object
+}
+
+
+publish_dataone <- function(in_file, out_file, code, meta, orcid){
+
+  dp <- new("DataPackage")
+  meta_obj <- data_object(meta, "eml://ecoinformatics.org/eml-2.2.0")
+  in_obj <- data_object(in_file)
+  code_obj <- data_object(code, mediaType="text/x-rsrc")
+  out_obj <- data_object(out_file)
+  dp <- addMember(dp, meta)
+  dp <- addMember(dp, in_obj, meta)
+  dp <- addMember(dp, code_obj, meta)
+  dp <- addMember(dp,  out_obj, meta)
+  rules <- data.frame(subject=orcid, permission="changePermission") 
```

Comment:
This might not be working. Our authentication system represents users with the `http` version of their ORCID URI -- not the `https` version -- and doesn't consider the two equivalent. I'd expect your rule here isn't working but it's not causing a problem because you are also what we call `rightsHolder` on the objects which gives you the same permission anyway. You could remove this line, put some code in here to make `subject` use the `http://orcid.org/...` variant of your ORCID, or wait for us to fix things on our end.

Reply:
(PS this all looks really nice and I think you've done a nice job keeping a streamlined approach here)

Reply:
Thanks much for taking a close look!  Is the https ORCID going to change on the DataONE end? (I think I saw some chatter about that -- I believe ORCID considers the https version as the canonical form these days?

Also looks like SHA-256 is now the default in `datapack` so we can streamline the above a bit more.

Reply:
I think we're hoping to change to using the `https` variant, yeah. So your code here, unchanged, will eventually work.